util/tracing/detect/detect_v2.go

```diff
@@ -0,0 +1,203 @@
+package detect
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/pkg/errors"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+)
+
+type exporterFactory struct {
+	TracerFactory TracerFactory
+	MetricFactory MetricFactory
+}
+
+type ExporterFactory[T any] interface {
+	// AutoConfigure returns true if this exporter is automatically configured
+	// with the current environment.
+	AutoConfigure() bool
+
+	// New constructs a new exporter of this type.
+	New() (T, error)
+}
+
+type (
+	TracerFactory = ExporterFactory[sdktrace.SpanProcessor]
+	MetricFactory = ExporterFactory[sdkmetric.Reader]
+)
+
+var exporters map[string]exporterFactory
+
+type ExporterFactoryOption interface {
+	apply(factory *exporterFactory)
+}
+
+type exporterFactoryOption func(factory *exporterFactory)
+
+func (fn exporterFactoryOption) apply(factory *exporterFactory) {
+	fn(factory)
+}
+
+func WithTracerFactory(factory TracerFactory) ExporterFactoryOption {
+	return exporterFactoryOption(func(ef *exporterFactory) {
+		ef.TracerFactory = factory
+	})
+}
+
+func WithMetricFactory(factory MetricFactory) ExporterFactoryOption {
+	return exporterFactoryOption(func(ef *exporterFactory) {
+		ef.MetricFactory = factory
+	})
+}
+
+func RegisterExporter(name string, opts ...ExporterFactoryOption) {
+	if exporters == nil {
+		exporters = make(map[string]exporterFactory)
+	}
+
+	ef := exporterFactory{}
+	for _, opt := range opts {
+		opt.apply(&ef)
+	}
+	exporters[name] = ef
+}
+
+func Resource(ctx context.Context) (*resource.Resource, error) {
+	res, err := resource.Detect(ctx, serviceNameDetector{})
```

Comment:
I _think_ `serviceNameDetector` can go away entirely as part of this refactor and `resource.Default()` can be used directly (discovered this with @thaJeztah @ https://github.com/moby/moby/pull/46830#discussion_r1422677354)

> It looks like `resource.Default()` is sufficient:
> * It uses `fromEnv{}` detector internally: https://github.com/open-telemetry/opentelemetry-go/blob/v1.19.0/sdk/resource/resource.go#L208
> * `fromEnv{}` detector reads `OTEL_SERVICE_NAME`: https://github.com/open-telemetry/opentelemetry-go/blob/v1.19.0/sdk/resource/env.go#L53

(`serviceNameDetector` is 2+ years old, so I'm guessing it was written it _before_ the SDK supported the standard env var, but maybe there's additional subtlety here I'm unaware of)


See also:
* moby/moby#46830

Reply:
The primary difference from what I understand is there's a slight difference in the logic for the default service name. For the environment variable, both will return the same thing. For the default name, you'll get this difference:

https://github.com/moby/buildkit/blob/5ecfec1039959e8b742494eec8f7b5fd8202df5a/vendor/go.opentelemetry.io/otel/sdk/resource/builtin.go#L101-L105

Will cause the default service name to be the executable path.

https://github.com/moby/buildkit/blob/f84cfe3d6ab7b0645df76e82a02458d3088f32a9/util/tracing/detect/detect.go#L228

This one uses the base path. There doesn't seem to be a good way to modify how `service.name` gets set. If you merge it into the default resource, you overwrite the auto-discovered name. If you merge the default resource into it, this default seems to get populated. Although we might be able to use `Default -> custom one -> Environment` and that would likely work. I'll see if I can rework it to remove the `serviceNameDetector`.

Reply:
It's definitely possible to change this.

In particular, `resource.New` and then using the detector options like `resource.WithFromEnv()` and `resource.WithTelemetrySDK()`.